### PR TITLE
FIN, RESET_STREAM, RESET_STREAM_AT, STOP_SENDING

### PR DESCRIPTION
**Pull request by martinduke**
This is explicitly not dealing with FETCH at all, but it won't be hard for me to write something similar for the FETCH stream once the rubble settles on that proposal.

### CODE REVIEW

File: draft-ietf-moq-transport.md [hunk 1]
```diff
@@ -1986,6 +1986,66 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+
+If a sender has delivered all objects in a Subgroup, except any objects that
+fall before the beginning of a subscription, it MUST close the stream
+with a FIN.
+
+If a sender closes the stream before verifying it has delivered all such
+objects when it , it MUST use a RESET_STREAM or RESET_STREAM_AT
+{{!I-D.draft-ietf-quic-reliable-stream-reset}} frame. This includes early
+termination of subscription due to an UNSUBSCRIBE message, a sender's
+decision to abandon the subscription before the Group is complete, or a
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:2005]: What if you give up on delivering a Subgroup due to Delivery Timeout?

For abandoning a subscription early, that corresponds to sending a SUBSCRIBE_DONE, correct?

File: draft-ietf-moq-transport.md [hunk 2]
```diff
@@ -1986,6 +1986,66 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+Group.
+
+If a sender does not deliver any objects from a Subgroup of a subscribed Group,
+it MAY send a STREAM_HEADER_SUBGROUP on a new data stream, with no objects, and
+then send RESET_STREAM_AT with a reliable_size equivalent to the length of the
+stream header. This would explicitly tell the receiver that there are unsent
+Subgroups in the Group.
+
+Since SUBSCRIBEs always end on a group boundary [TODO: Update SUBSCRIBE to
+express this], an ending subscription can always cleanly close all its
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:2016]: Should we do that change in a separate PR and land it first?

File: draft-ietf-moq-transport.md [hunk 3]
```diff
@@ -1986,6 +1986,66 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+subgroups. A sender that terminates a stream early for any other reason (e.g.,
+to handoff to a different sender MUST use RESET_STREAM or RESET_STREAM_AT.
+Senders SHOULD terminate a stream on Group boundaries to avoid doing so.
+
+An Original Publisher MAY create EndOfSubgroup objects to provide further
+indication that streams are OK to close.
+
+[TODO: Do we still need EndOfSubgroup objects?]
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:2024]:
```suggestion
[TODO: Do we still need EndOfSubgroup objects?]
```

Good question, I don't think so?  Fetch doesn't need them clearly.

File: draft-ietf-moq-transport.md [hunk 4]
```diff
@@ -1986,6 +1986,66 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+
+Processing a RESET_STREAM or RESET_STREAM_AT means that there might be other
+objects in the Subgroup beyond the last one received. A relay might immediately
+reset the corresponding downstream stream, or it might attempt to recover the
+missing Objects in an effort to send FIN, presumably with some timeout. It also
+might send RESET_STREAM_AT with reliable_size set to the last object it has, so
+as to reliably deliver the objects it has while signaling that other objects
+exist.
+
+A relay MAY also use EndOfGroup messages in a Subgroup stream to close streams
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:2041]: Can we just remove this, as it is implied, but seems complex.

File: draft-ietf-moq-transport.md [hunk 5]
```diff
@@ -1986,6 +1986,66 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+for other Subgroups in that Group, if it has received all Object IDs between the
+highest Object ID in the subject Subgroup and the EndOfGroup object. This may be
+complex to implement.
+
+A receiver MAY send a QUIC STOP_SENDING frame on a subgroup stream if the Group
+is no longer of interest to it.
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:2047]:
```suggestion
A receiver MAY send a QUIC STOP_SENDING frame for a subgroup stream if the Group
or subgroup is no longer of interest to it.
```

**Review by ianswett — COMMENTED** [review]
LG overall, some small comments.

File: draft-ietf-moq-transport.md (hunk 1 above)
Comment (martinduke) [reply draft-ietf-moq-transport.md:2005]: I made Delivery Timeout as a reason explicit, instead of implicit.

As another issue makes clear, I don't know what SUBSCRIBE_DONE means. Is it a "Reset" or a "Fin" of a subscription? Or both?

File: draft-ietf-moq-transport.md (hunk 2 above)
Comment (martinduke) [reply draft-ietf-moq-transport.md:2016]: There's a TODO there to give you flexibility on ordering, but do what you like.

File: draft-ietf-moq-transport.md (hunk 3 above)
Comment (martinduke) [reply draft-ietf-moq-transport.md:2024]: I would like to get rid of these, but am happy to do it in a different PR to avoid controversy. If no one wants EndOf Subgroup, we can kill it now.

The edge case I can imagine is if FETCH contents can be SUBSCRIBEd (see #588). There's some ambiguity about FIN if the FETCH ends in the middle of the group. I would be happy to have that uncertainty, personally, but others might disagree.

File: draft-ietf-moq-transport.md (hunk 4 above)
Comment (martinduke) [reply draft-ietf-moq-transport.md:2041]: I won't die on this hill, but why not explicitly give people permission to do this?

File: draft-ietf-moq-transport.md [hunk 6]
```diff
@@ -2183,6 +2183,65 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
 
+### Closing Subgroup Streams
+
+Subscribers will often need to know if they have received all objects in a
+Subgroup, particularly if they serve as a relay or cache. QUIC and Webtransport
+streams provide signals that can be used for this purpose. Closing Subgroups
+promptly frees system resources and often unlocks flow control credit to open
+more streams.
+
+If a sender has delivered all objects in a Subgroup, except any objects that
```
Comment (afrind) [inline draft-ietf-moq-transport.md:2194]: > a sender has delivered all objects in a Subgroup

Seems like this only applies to the original publisher, right?  For a relay, you need to explain how they know they have delivered all the objects in a subgroup.

Also, what does "delivered" mean?  I think you mean they've been committed to the QUIC stream, rather than transmitted or acknowledged.

Should we say "requested" objects to be inclusive of subscribe filters?

File: draft-ietf-moq-transport.md [hunk 7]
```diff
@@ -2183,6 +2183,65 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+fall before the beginning of a subscription, it MUST close the stream
+with a FIN.
+
+If a sender closes the stream before verifying it has delivered all such
```
Comment (afrind) [inline draft-ietf-moq-transport.md:2198]: Same comment here regarding "verifying it has delivered".

File: draft-ietf-moq-transport.md [hunk 8]
```diff
@@ -2183,6 +2183,65 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+objects, it MUST use a RESET_STREAM or RESET_STREAM_AT
+{{!I-D.draft-ietf-quic-reliable-stream-reset}} frame. This includes an open
+Group exceeding its Delivery Timeout, early termination of subscription due to
```
Comment (afrind) [inline draft-ietf-moq-transport.md:2201]: s/Group/Subgroup/ throughout this paragraph?

File: draft-ietf-moq-transport.md (hunk 3 above)
Comment (afrind) [reply draft-ietf-moq-transport.md:2221]: I'd like to get rid of them too.  Setting the object ID is problematic.  If we need to, we can retain them for FETCH specifically (though the object ID problem is still there.

File: draft-ietf-moq-transport.md [hunk 9]
```diff
@@ -2183,6 +2183,65 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+to handoff to a different sender MUST use RESET_STREAM or RESET_STREAM_AT.
+Senders SHOULD terminate a stream on Group boundaries to avoid doing so.
+
+An Original Publisher MAY create EndOfSubgroup objects to provide further
+indication that streams are OK to close.
+
+[TODO: Do we still need EndOfSubgroup objects?]
+
+An MoQT implementation that processes a stream FIN is assured it has received
+all objects in a subgroup from the start of the subscription. If a relay, it
```
Comment (afrind) [inline draft-ietf-moq-transport.md:2224]: The indicates that delivery timeout *does not* allow skipping objects in the middle of the subgroup.  Is the group aligned here?

File: draft-ietf-moq-transport.md [hunk 10]
```diff
@@ -2183,6 +2183,65 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+can forward stream FINs to its own subscribers once those objects have been
+sent. A relay MAY use EndOfGroup, EndOfSubgroup, GroupDoesNotExist, and
```
Comment (afrind) [inline draft-ietf-moq-transport.md:2226]: Maybe add "received" in there to show the relay is not injecting these objects into the stream.

File: draft-ietf-moq-transport.md [hunk 11]
```diff
@@ -2183,6 +2183,65 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+as to reliably deliver the objects it has while signaling that other objects
+exist.
+
+A relay MAY also use EndOfGroup messages in a Subgroup stream to close streams
+for other Subgroups in that Group, if it has received all Object IDs between the
+highest Object ID in the subject Subgroup and the EndOfGroup object. This may be
+complex to implement.
+
+A receiver MAY send a QUIC STOP_SENDING frame for a subgroup stream if the Group
+or subgroup is no longer of interest to it.
```
Comment (afrind) [inline draft-ietf-moq-transport.md:2244]: Should we explain what the publisher does when it receives STOP_SENDING?

**Review by afrind — COMMENTED** [review]
Individual Review

File: draft-ietf-moq-transport.md (hunk 9 above)
Comment (martinduke) [reply draft-ietf-moq-transport.md:2231]: it better be! subgroups imply dependency, so it is pointless to skip objects.

File: draft-ietf-moq-transport.md (hunk 10 above)
Comment (martinduke) [reply draft-ietf-moq-transport.md:2233]: sure

File: draft-ietf-moq-transport.md (hunk 11 above)
Comment (martinduke) [reply draft-ietf-moq-transport.md:2251]: I think RFC9000 is pretty clear that STOP_SENDING -> RESET_STREAM.

File: draft-ietf-moq-transport.md (hunk 7 above)
Comment (martinduke) [reply draft-ietf-moq-transport.md]: OK, I changed this to enumerate three cases:

sent everything and wait for reliable delivery: FIN

didn't send everything: RESET_STREAM

sent everything, don't wait for delivery: FIN followed by RESET_STREAM.

File: draft-ietf-moq-transport.md (hunk 11 above)
Comment (afrind) [reply draft-ietf-moq-transport.md:2251]: I think some explanatory text about RESET_STREAM vs RESET_STREAM_AT, and also indicating that this shouldn't impact the rest of the subscription would be helpful.

File: draft-ietf-moq-transport.md (hunk 9 above)
Comment (afrind) [reply draft-ietf-moq-transport.md:2231]: See discussion here: https://github.com/moq-wg/moq-transport/pull/587#discussion_r1799916677

**Review by ianswett — COMMENTED** [review]
Some editorial comments, but this still LG from a functional perspective.

File: draft-ietf-moq-transport.md (hunk 3 above)
Comment (martinduke) [reply draft-ietf-moq-transport.md:2228]: it's not that there's a "problem" with Fetch; it's that if there are EndOfSubgroup objects, you are potentially sending a LOT of them in response to a FETCH.

Not in this PR, but I think we should get rid of EndOfSubgroup entirely. FETCH response should omit DNE and EndOfGroup status, because they are implied by sequence number gaps and the beginning of the next group, respectively.

File: draft-ietf-moq-transport.md (hunk 4 above)
Comment (martinduke) [reply draft-ietf-moq-transport.md:2245]: I slightly reworded and moved this paragraph, but I don't know if that solved your problem.

File: draft-ietf-moq-transport.md [hunk 12]
```diff
@@ -2183,6 +2183,72 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+an UNSUBSCRIBE message, a sender's decision to abandon the subscription before
+the Subgroup is complete, or a SUBSCRIBE_UPDATE moving the end of the subscription
+to before the current Group.
+
+A sender might deliver all objects in a Subgroup to the QUIC stream, including a
+FIN, and then reset the stream because any retransmissions are unnecessary due
+to any of the reset conditions above being met. In this case, the receiving
+application would receive the FIN if and only if all objects were, in fact,
+delivered. The application can therefore ignore a RESET_STREAM if it receives
+FIN.
```
[the remark this answers is not in the document]
Comment (martinduke) [reply draft-ietf-moq-transport.md:2210]: All objects were received, not sent. Accepting, then further editing.

File: draft-ietf-moq-transport.md (hunk 4 above)
Comment (ianswett) [reply draft-ietf-moq-transport.md]: I did make a suggestion on a way to rewrite the paragraph

File: draft-ietf-moq-transport.md [hunk 13]
```diff
@@ -2183,6 +2183,72 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+reset the corresponding downstream stream, or it might attempt to recover the
+missing Objects in an effort to send FIN, presumably with some timeout. It also
+might send RESET_STREAM_AT with reliable_size set to the last object it has, so
+as to reliably deliver the objects it has while signaling that other objects
+exist.
+
+A relay MAY also use EndOfGroup messages in a Subgroup stream to close streams
+for other Subgroups in that Group, if it has received all Object IDs between the
+highest Object ID in the subject Subgroup and the EndOfGroup object. This may be
+complex to implement.
```
[the remark this answers is not in the document]
Comment (martinduke) [reply draft-ietf-moq-transport.md]: the Github thread is weird so I didn't see this until I'd already edited, but I manually copied in this wording.

File: draft-ietf-moq-transport.md [hunk 14]
```diff
@@ -2183,6 +2183,72 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+indication that streams are OK to close.
+
+[TODO: Do we still need EndOfSubgroup objects?]
+
+An MoQT implementation that processes a stream FIN is assured it has received
+all objects in a subgroup from the start of the subscription. If a relay, it
+can forward stream FINs to its own subscribers once those objects have been
+sent. A relay MAY use EndOfGroup, EndOfSubgroup, GroupDoesNotExist, and
+EndOfTrack objects to close streams even if the FIN has not arrived, as further
+objects on the stream would be a protocol violation.
```
[the remark this answers is not in the document]
Comment (martinduke) [reply draft-ietf-moq-transport.md:2231]: Sometimes the FIN arrives in a separate QUIC packet. This sentence allows you to anticipate the later FIN.

File: draft-ietf-moq-transport.md (hunk 2 above)
Comment (ianswett) [reply draft-ietf-moq-transport.md:2216]: See #593

File: draft-ietf-moq-transport.md [hunk 15]
```diff
@@ -2183,6 +2183,75 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
```
Comment (afrind) [inline draft-ietf-moq-transport.md:2231]:
```suggestion
sent. A relay MAY treat receipt of EndOfGroup, EndOfSubgroup, GroupDoesNotExist, or
EndOfTrack objects as a signal to close corresponding streams even if the FIN has not arrived, as further
objects on the stream would be a protocol violation.
```

File: draft-ietf-moq-transport.md [hunk 16]
```diff
@@ -2183,6 +2183,75 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+
+Similarly, an EndOfGroup message indicates the total number of Objects in the
```
Comment (afrind) [inline draft-ietf-moq-transport.md:2233]: This is only true if Object IDs are mandated to be sequential within a group, which I don't think is a restriction.

File: draft-ietf-moq-transport.md [hunk 17]
```diff
@@ -2183,6 +2183,75 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+any stream where the entire subgroup has been sent. This might be complex to
+implement.
+
+Processing a RESET_STREAM or RESET_STREAM_AT means that there might be other
+objects in the Subgroup beyond the last one received. A relay might immediately
+reset the corresponding downstream stream, or it might attempt to recover the
+missing Objects in an effort send all the objects in the subgroups and the FIN. It also
+might send RESET_STREAM_AT with reliable_size set to the last object it has, so
+as to reliably deliver the objects it has while signaling that other objects
+exist.
```
Comment (afrind) [inline draft-ietf-moq-transport.md:2244]:
```suggestion
as to reliably deliver the objects it has while signaling that other objects
might exist.
```

File: draft-ietf-moq-transport.md (hunk 16 above)
Comment (martinduke) [reply draft-ietf-moq-transport.md:2233]: Please refer to last Wednesday's slides! There are times you can do this if you make your implementation complex enough.

Comment (martinduke) [reply draft-ietf-moq-transport.md]: ah, he's referring to number of objects vs max object iD. Fixed.

File: draft-ietf-moq-transport.md (hunk 1 above)
Comment (ianswett) [reply draft-ietf-moq-transport.md]: It's typically a 'Fin' of a subscription.  Maybe one sends a SUBSCRIBE_DONE when all subgroups have been Fin'd?

File: draft-ietf-moq-transport.md (hunk 3 above)
Comment (ianswett) [reply draft-ietf-moq-transport.md:2224]: I wrote #596 to remove End Of Subgroup

File: draft-ietf-moq-transport.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (ianswett) [reply draft-ietf-moq-transport.md:2231]: I'm still confused why a sender would do this, but I agree it's probably ok to do this.

File: draft-ietf-moq-transport.md (hunk 1 above)
Comment (martinduke) [reply draft-ietf-moq-transport.md]: Well if it's a "subscribe fin", then it ends on a group boundary and all the streams will naturally close with their own FIN, with no special mechanism required.

File: draft-ietf-moq-transport.md [hunk 18]
```diff
@@ -2175,6 +2175,70 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+
+A receiver MAY send a QUIC STOP_SENDING frame for a subgroup stream if the Group
+or Subgroup is no longer of interest to it. The sender SHOULD respond with
+RESET_STREAM or RESET_STREAM_AT. If RESET_STREAM_AT, note that the receiver has
+indicated no interest in the objects, so setting a reliable_size beyond the
+stream header is of questionable utility.
+
+RESET_STREAM and STOP_SENDING on SUBSCRIBE data streams have no impact on other
+Subgroups in the Group or the subscription at large, although in practice all
+Subgroups in a Group will sometimes be canceled at once.
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:2241]:
```suggestion
Subgroups in the Group or the subscription, although applications might cancel all
Subgroups in a Group at once.
```

File: draft-ietf-moq-transport.md [hunk 19]
```diff
@@ -2175,6 +2175,70 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+
+If a sender has delivered all objects in a Subgroup to the QUIC stream, except
+any objects before the beginning of a subscription, it MUST close the
+stream with a FIN.
+
+If a sender closes the stream before delivering all such objects to the QUIC
+stream, it MUST use a RESET_STREAM or RESET_STREAM_AT
+{{!I-D.draft-ietf-quic-reliable-stream-reset}} frame. This includes an open
+Subgroup exceeding its Delivery Timeout, early termination of subscription due to
+an UNSUBSCRIBE message, a sender's decision to abandon the subscription, or a
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:2194]:
```suggestion
an UNSUBSCRIBE message, a publisher's decision to end the subscription early, or a
```

File: draft-ietf-moq-transport.md [hunk 20]
```diff
@@ -2175,6 +2175,70 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+If a sender has delivered all objects in a Subgroup to the QUIC stream, except
+any objects before the beginning of a subscription, it MUST close the
+stream with a FIN.
+
+If a sender closes the stream before delivering all such objects to the QUIC
+stream, it MUST use a RESET_STREAM or RESET_STREAM_AT
+{{!I-D.draft-ietf-quic-reliable-stream-reset}} frame. This includes an open
+Subgroup exceeding its Delivery Timeout, early termination of subscription due to
+an UNSUBSCRIBE message, a publisher's decision to end the subscription early, or a
+SUBSCRIBE_UPDATE moving the end of the subscription to before the current Group.
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:2195]:
```suggestion
SUBSCRIBE_UPDATE moving the end of the subscription to before the current Group
or the start after the current Group.
```

**Review by ianswett — APPROVED** [review]
I've tried to make any editorial changes so this is ready for merging I believe.

File: draft-ietf-moq-transport.md (hunk 9 above)
Comment (afrind) [reply draft-ietf-moq-transport.md:2215]: This was discussed 10/23.  Cullen agreed with this interpretation.

File: draft-ietf-moq-transport.md [hunk 21]
```diff
@@ -2175,6 +2175,71 @@ The Object Status field is only sent if the Object Payload Length is zero.
...
+Group boundaries to avoid doing so.
+
+An MoQT implementation that processes a stream FIN is assured it has received
+all objects in a subgroup from the start of the subscription. If a relay, it
+can forward stream FINs to its own subscribers once those objects have been
+sent. A relay MAY treat receipt of EndOfGroup, EndOfSubgroup, GroupDoesNotExist, or
+EndOfTrack objects as a signal to close corresponding streams even if the FIN
+has not arrived, as further objects on the stream would be a protocol violation.
+
+Similarly, an EndOfGroup message indicates the total maximum Object ID in the
```
Comment (ianswett) [inline draft-ietf-moq-transport.md:2221]:
```suggestion
Similarly, an EndOfGroup message indicates the maximum Object ID in the
```

**Comment by martinduke** [comment]
> I've tried to make any editorial changes so this is ready for merging I believe.

LGTM
